cmd/manager/aggregator.go

```diff
@@ -360,88 +358,96 @@ func createResults(crClient *complianceCrClient, scan *compv1alpha1.ComplianceSc
+			rem := pr.Remediations[idx]
+			if remErr := handleRemediation(crClient, rem, pr.CheckResult, scan); remErr != nil {
+				return remErr
 			}
-			continue
 		}
+	}
 
-		remLabels := getRemediationLabels(scan, remTargetObj)
+	return nil
+}
 
-		// The state even if set in the object would have been overwritten by the call to
-		// spec update, so we keep the state separately in a variable
-		stateUpdate := compv1alpha1.RemediationPending
+func handleRemediation(crClient *complianceCrClient, rem *compv1alpha1.ComplianceRemediation, cr *compv1alpha1.ComplianceCheckResult, scan *compv1alpha1.ComplianceScan) error {
+	crkey := getObjKey(cr.GetName(), cr.GetNamespace())
+	remTargetObj := rem.Spec.Current.Object
 
-		remkey := getObjKey(pr.Remediation.GetName(), pr.Remediation.GetNamespace())
-		foundRemediation := &compv1alpha1.ComplianceRemediation{}
-		// Copy type metadata so dynamic client copies data correctly
-		foundRemediation.TypeMeta = pr.Remediation.TypeMeta
-		log.Info("Getting ComplianceRemediation", "ComplianceRemediation.Name", crkey.Name,
-			"ComplianceRemediation.Namespace", crkey.Namespace)
-		remExists := getObjectIfFound(crClient, remkey, foundRemediation)
-		if remExists {
-			// If the remediation is already applied and the status of the check is compliant, only update
-			// the remediation if the payload differs. Let's not create remediations for checks that are passing
-			// needlessly and let's not trigger the remediation controller needlessly
-			if foundRemediation.Status.ApplicationState == compv1alpha1.RemediationApplied ||
-				foundRemediation.Status.ApplicationState == compv1alpha1.RemediationOutdated {
-				if !foundRemediation.RemediationPayloadDiffers(pr.Remediation) {
-					log.Info("Not updating passing remediation that was the same between runs", "ComplianceRemediation.Name", foundRemediation.Name)
-					continue
-				}
-
-				// Applied remediation that differs must be updated, let's set the appropriate state
-				stateUpdate = compv1alpha1.RemediationOutdated
-				if foundRemediation.Status.ApplicationState == compv1alpha1.RemediationApplied {
-					// For applied remediations, the old state must be kept in the outdated field
-					// so that the admin can switch to the current state at their own pace
-					foundRemediation.Spec.Current.DeepCopyInto(&pr.Remediation.Spec.Outdated)
-				} else {
-					// For remediations that were already outdated, keep the outdated data there
-					foundRemediation.Spec.Outdated.DeepCopyInto(&pr.Remediation.Spec.Outdated)
-				}
-				// The application state must be preserved or else we'd un-apply the remediation
-				// once it's updated
-				pr.Remediation.Spec.Apply = foundRemediation.Spec.Apply
-				// Also label the outdated remediations so that the admin can find them
-				remLabels[compv1alpha1.OutdatedRemediationLabel] = ""
-			}
+	if canCreate, why := canCreateRemediation(scan, remTargetObj); !canCreate {
+		// Only issue event once.
```

Comment:
The comment still says the event is issued once, while it's not. We should either remove the comment or keep the boolean that adds the guard back.

Reply:
I removed the comment.